src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -152,12 +156,83 @@ class SetupPyChroot:
   setup_keywords_json: str
 
 
+@dataclass(frozen=True)
+class RunSetupPyRequest:
+  """A request to run a setup.py command."""
+  exported_target: ExportedTarget
+  chroot: SetupPyChroot
+  args: Tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class RunSetupPyResult:
+  """The result of running a setup.py command."""
+  output: Digest  # The state of the chroot after running setup.py.
+
+
 @dataclass(frozen=True)
 class SetuptoolsSetup:
   """The setuptools tool."""
   requirements_pex: Pex
 
 
+class SetupPyOptions(GoalSubsystem):
+  name = "setup-py2"
+
+  @classmethod
+  def register_options(cls, register):
+    super().register_options(register)
+    register(
+      '--args', type=list, member_type=str,
+      help="Arguments to pass directly to setup.py, e.g. "
+           "`--setup-py2-args=\"bdist_wheel\"`. If unspecified, we just dump the setup.py chroot."
+    )
+
+
+class SetupPy(Goal):
+  """Runs setup.py commands."""
+  subsystem_cls = SetupPyOptions
+
+
+@console_rule
+async def run_setup_pys(addresses: BuildFileAddresses, console: Console, options: SetupPyOptions,
+                        distdir: DistDir, workspace: Workspace) -> SetupPy:
+  """Run setup.py commands on all exported targets addressed."""
+  targets = await Get[HydratedTargets](BuildFileAddresses, addresses)
+
+  exported_targets: List[ExportedTarget] = []
+  nonexported_targets: List[HydratedTarget] = []
+  for target in targets:
+    if _is_exported(target):
+      exported_targets.append(ExportedTarget(target))
+    else:
+      nonexported_targets.append(target)
+  if nonexported_targets:
+    raise TargetNotExported(
+      'Cannot run setup.py on these targets, because they have no `provides=` clause: '
+      f'{", ".join(so.address.reference() for so in nonexported_targets)}')
+
+  for target in exported_targets:
+    addr = target.hydrated_target.address.reference()
+    chroot = await Get[SetupPyChroot](SetupPyChrootRequest(target))
+    provides = target.hydrated_target.adaptor.provides
+    setup_py_dir = distdir.relpath / f'{provides.name}-{provides.version}'
+    if options.values.args:
+      console.print_stderr(f'Running setup.py for {addr}')
+      setup_py_result = await Get[RunSetupPyResult](
+        RunSetupPyRequest(target, chroot, tuple(options.values.args)))
```

Comment:
This won't work properly with shlexing. Instead, define on the `GoalSubsystem` a function `get_args()` that does this:

https://github.com/pantsbuild/pants/blob/a6716b441e260a5c8cce663d4c4df76c4c4d7a14/src/python/pants/backend/python/lint/black/subsystem.py#L31-L32

Fwit, the refactoring I've been doing on `option/` the past few days has been prework to move support for shlexable strings directly into the parser itself to avoid this "gotcha".

Reply:
Once we merge https://github.com/pantsbuild/pants/pull/8904, this will be irrelevant. I'm planning on merging once I get one more review, so probably not worth your time to add `get_args()` and then have to delete it tomorrow.

Reply:
Done! And I stand chastised by my falling for the gotcha I termed "minor"...